hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/DispatchProcessor.java

```diff
@@ -148,15 +148,9 @@ private void tryHandle(
             }
             handleSystemUpdates(dispatch);
         } catch (HandleException e) {
-            logger.debug("Transaction failed handle", e);
-            // In case of a ContractCall when it reverts, the gas charged should not be rolled back
-            rollback(e.shouldRollbackStack(), e.getStatus(), dispatch.stack(), dispatch.streamBuilder());
-            if (e.shouldRollbackStack()) {
-                chargePayer(dispatch, validation, false);
-                e.maybeReplayFees(dispatch);
-            }
-            // Since there is no easy way to say how much work was done in the failed dispatch,
-            // and current throttling is very rough-grained, we just return USER_TRANSACTION here
+            rollback(e.getStatus(), dispatch.stack(), dispatch.streamBuilder());
+            chargePayer(dispatch, validation, false);
+            e.maybeReplayFees(dispatch);
```

Comment:
A question just for my understanding.  Why is this replay necessary?  It seems that the actual fees are paid in the line above and that the replay does not change the amount charged.  Is this so that the record is correct?

Reply:
The `chargePayer(...)` call **only** re-charges the fees the workflow can "see", based on what was returned by the dispatch's `TransactionHander#calculateFees()` method.

But we want to **also** re-charge for gas. This is what `e.maybeReplayFees(...)` accomplishes.